vispy/visuals/mesh_normals.py

```diff
@@ -76,14 +76,59 @@ class MeshNormalsVisual(LineVisual):
 
     def __init__(self, meshdata, primitive='face', length=None,
                  length_method='median_edge', length_scale=1.0, **kwargs):
+        self._previous_meshdata = None
+        super().__init__(connect='segments')
+        self.set_data(meshdata, primitive, length, length_method, length_scale, **kwargs)
+
+    def set_data(self, meshdata=None, primitive='face', length=None,
+                 length_method='median_edge', length_scale=1.0, **kwargs):
+        """Set the data used to draw this visual
+
+        Parameters
+        ----------
+        meshdata : instance of :class:`~vispy.geometry.meshdata.MeshData`
+            The mesh data.
+        primitive : {'face', 'vertex'}
+            The primitive type on which to compute and display the normals.
+        length : None or float or array-like, optional
+            The length(s) of the normals. If None, the length is computed with
+            `length_method`.
+        length_method : {'median_edge', 'max_extent'}, default='median_edge'
+            The method to compute the length of the normals (when `length=None`).
+            Methods: 'median_edge', the median edge length; 'max_extent', the
+            maximum side length of the bounding box of the mesh.
+        length_scale : float, default=1.0
+            A scale factor applied to the length computed with `length_method`.
+        **kwargs : dict, optional
+            Extra arguments to define the appearance of lines. Refer to
+            :class:`~vispy.visuals.line.line.LineVisual`.
+        """
         if primitive not in ('face', 'vertex'):
             raise ValueError('primitive must be "face" or "vertex", got %s'
                              % primitive)
+        # remove connect from kwargs to make sure we don't change it
+        kwargs.pop('connect', None)
+
+        if meshdata is None:
+            meshdata = self._previous_meshdata
 
         if primitive == 'face':
-            normals = meshdata.get_face_normals()
+            if meshdata._faces is None:
```

Comment:
Could you change this to `meshdata.get_faces()` and similarly the vertices check below? Avoid using the private API. Or would it be good to have a `__nonzero__` or `is_empty` method added to MeshData and then you could check that before these face/vertex if statements.

Additionally, could you add an `else:` (or an extra check above) for the case where `primitive` is not face or vertex. As is right now `normals` could not exist by the time it gets to `if normals is None` below.

Reply:
That can be done for faces, but `get_vertices` computes some stuff and that was at the root of some of the issues.

Reply:
The check for `primitive` already exists above!

Reply:
What do you think about the is_empty-like checks on the MeshData itself? Something to hide the private property checks.

Reply:
Sounds reasonable!